### DISCUSSION TITLE
PKCS#11 signature example. Hardware eToken 5100

### DISCUSSION THREAD

**Discussion by C-monC** [MatthiasValvekens/pyHanko#308] [Q&A] [answered]
Hi, I am struggling to put together a bare minimum PKCS11 sign script.
```
from pyhanko import stamp
from pyhanko.pdf_utils import images
from pyhanko.pdf_utils.incremental_writer import IncrementalPdfFileWriter
from pyhanko.sign import fields, signers
from pyhanko.sign.pkcs11 import PKCS11Signer, PKCS11SigningContext
from pyhanko.config.pkcs11 import PKCS11SignatureConfig

conf = PKCS11SignatureConfig(
    module_path="/usr/lib/libeTPkcs11.so",
    cert_label="token",
    user_pin="12345"
)
ctx = PKCS11SigningContext(conf)

signer = PKCS11Signer()
```
How do I pass the config to the signer? The only function/class that I found that consumes the config is PKCS11SigningContext but I have no idea what to do with it. 

Here is the output from pkcs11-tool

`pkcs11-tool --modul /usr/lib/libeTPkcs11.so --show-info -T`

> Cryptoki version 2.20
> Manufacturer     SafeNet, Inc.
> Library          SafeNet eToken PKCS#11 (ver 10.8)
> Available slots:
> Slot 0 (0x0): SafeNet eToken 5100 [eToken 5110 SC] 00 00
>   token label        : token
>   token manufacturer : SafeNet, Inc.
>   token model        : eToken
>   token flags        : login required, rng, token initialized, PIN initialized, other flags=0x200
>   hardware version   : 16.0
>   firmware version   : 0.0
>   serial num         : 123455678
>   pin min/max        : 8/20
> Using slot 0 with a present token (0x0)

**Comment by MatthiasValvekens** [comment]
This should probably be clarified in the docs, but `PKCS11SigningContext` is a Python context manager that exposes a pyHanko `Signer`. It's done that way to ensure the PKCS#11 session is opened/closed at the appropriate time. A minimal usage example looks like this:

```
    config = PKCS11SignatureConfig(...)
    w = IncrementalPdfFileWriter(BytesIO(MINIMAL))
    settings = signers.PdfSignatureMetadata(field_name='Sig1', ...)

    with PKCS11SigningContext(config) as signer:
        out = signers.sign_pdf(w, settings, signer=signer)

   # do something with the signed output...
```

I'll review the library docs and try to find a good place to add an example.

**Comment by C-monC** [comment]
Thanks for the tips. I am still struggling to sign but I think it is a new issue. I see pyhanko looks for a Key where the class is PRIVATE_KEY, is this not blocked by most hardware tokens? My basic understanding of these hardware keys is that the private key is not accessible 

```
Exception has occurred: NoSuchKey
No key matching {<Attribute.CLASS>: <ObjectClass.PRIVATE_KEY>, <Attribute.LABEL>: '[cert-label]'}
StopIteration: 
```
The code that causes this error:
```
conf = PKCS11SignatureConfig(
    # slot_no=0,
    module_path="/usr/lib/libeTPkcs11.so",
    cert_label="[cert-label]",
    user_pin='pin-code',
    prompt_pin=PKCS11PinEntryMode.PROMPT,
    token_criteria=TokenCriteria(
        label='my_token',
        serial=b'token-serial-number'
    )
)


with open('sample1.pdf', 'rb') as inf:
    w = IncrementalPdfFileWriter(inf)
    fields.append_signature_field(
        w, sig_field_spec=fields.SigFieldSpec(
            'Signature', box=(200, 600, 400, 660)
        )
    )

    settings = signers.PdfSignatureMetadata(field_name='Signature')

    with open('document-signed.pdf', 'wb') as outf: 
        with PKCS11SigningContext(conf) as signer:
            out = signers.sign_pdf(outf, settings, signer=signer)

```
Using just the pkcs11 module I can start an authenticated session in on the token
```
import pkcs11
lib = pkcs11.lib('/usr/lib/libeTPkcs11.so')

token = lib.get_token(token_label='my_token')
with token.open(user_pin='pin_code') as session:
    print(session)
```

output of list objects
```
pkcs11-tool --modul /usr/lib/libeTPkcs11.so --list-objects
Using slot 0 with a present token (0x0)
Certificate Object; type = X.509 cert
  label:      [cert-label]
  subject:    DN: ......
  ID:         [cert-id]
Certificate Object; type = X.509 cert
  label:      
  subject:    DN: C=US, O=Entrust, Inc., OU=See www.entrust.net\/legal-terms, OU=(c) 2015 Entrust, Inc. - for authorized use only, CN=Entrust Class 3 Client CA - SHA256
Certificate Object; type = X.509 cert
  label:      
  subject:    DN: O=Entrust.net, OU=www.entrust.net\/CPS_2048 incorp. by ref. (limits liab.), OU=(c) 1999 Entrust.net Limited, CN=Entrust.net Certification Authority (2048)
```
output of pkcs11-tool -T
```
pkcs11-tool --modul /usr/lib/libeTPkcs11.so -T
Available slots:
Slot 0 (0x0): SafeNet eToken 5100 [eToken 5110 SC] 00 00
  token label        : my_token
  token manufacturer : SafeNet, Inc.
  token model        : eToken
  token flags        : login required, rng, token initialized, PIN initialized, other flags=0x200
  hardware version   : 16.0
  firmware version   : 0.0
  serial num         : token-serial-number
  pin min/max        : 8/20
```

**Reply by MatthiasValvekens** [reply]
> I see pyhanko looks for a Key where the class is PRIVATE_KEY, is this not blocked by most hardware tokens? My basic understanding of these hardware keys is that the private key is not accessible

Well, obviously it's not possible to extract the _value_ of the private key from the token, since (as you correctly point out) that would defeat the purpose of using secure hardware tokens in the first place. However, in order to perform signing operations, it's necessary to acquire a "handle" on the private key, which is what pyHanko is trying to do here. Just having access to the certificate is not enough.

Are you sure the label of the key is the same as the one of the certificate? This isn't the case for all tokens, FWIW. It could also be that `--list-tokens` doesn't show the private key unless you have `pkcs11-tool` log in with your PIN first.

**Comment by C-monC** [comment]
I ran the list tokens with a logged in session. It seems like the key label is empty:

`pkcs11-tool --module /usr/lib/libeTPkcs11.so --login --login-type user --list-objects`
```
Using slot 0 with a present token (0x0)
Logging in to "my_token".
Please enter User PIN: 
Private Key Object; RSA 
  label:      
  ID:         [cert-id]
  Usage:      decrypt, sign, unwrap
  Access:     sensitive, always sensitive, never extractable, local
Certificate Object; type = X.509 cert
  label:     [cert-label]
  subject:    DN:
  ID:         [cert-id]
Certificate Object; type = X.509 cert
  label:      
  subject:    DN: C=US, O=Entrust, Inc., OU=See www.entrust.net\/legal-terms, OU=(c) 2015 Entrust, Inc. - for authorized use only, CN=Entrust Class 3 Client CA - SHA256
Certificate Object; type = X.509 cert
  label:      
  subject:    DN: O=Entrust.net, OU=www.entrust.net\/CPS_2048 incorp. by ref. (limits liab.), OU=(c) 1999 Entrust.net Limited, CN=Entrust.net Certification Authority (2048)

```
I now modified the script to use the key_id.

```
conf = PKCS11SignatureConfig(
    slot_no=0,
    module_path="/usr/lib/libeTPkcs11.so",
    cert_label="[cert-label]",
    cert_id='[cert-id]',
    key_id='[cert-id]',
    user_pin='my_pin',
    prompt_pin=PKCS11PinEntryMode.PROMPT,
    token_criteria=TokenCriteria(
        label='my_token',
        serial=b'token-serial-number'
    )
)
```
Exception thrown:
> venv/lib/python3.10/site-packages/pyhanko/sign/pkcs11.py", line 410, in _pull_cert
>     q = pkcs11_session.get_objects(query_params)
>   File "pkcs11/_pkcs11.pyx", line 369, in pkcs11._pkcs11.Session.get_objects
>   File "pkcs11/_pkcs11.pyx", line 294, in pkcs11._pkcs11.SearchIter.__init__
>   File "pkcs11/_pkcs11.pyx", line 69, in pkcs11._pkcs11.AttributeList.__cinit__
>   File "pkcs11/_utils.pyx", line 24, in pkcs11._pkcs11._pack_attribute
> TypeError: string argument without an encoding

Is it possible that the label is required?

**Reply by MatthiasValvekens** [reply, marked as answer]
Can't check now, but the key ID is probably supposed to be binary, not a text string.

**Reply by C-monC** [reply]
It worked! Thanks so much for your time.

```
from pkcs11 import Attribute, ObjectClass

lib = pkcs11.lib('/usr/lib/libeTPkcs11.so')

token = lib.get_token(token_label='my_token')
with token.open(user_pin='my_pin') as session:
    print(session)
    for obj in session.get_objects({
    Attribute.CLASS: ObjectClass.PRIVATE_KEY,
    }): 
        print(obj) <--- break point here check id
```

This gave me the key's ID in this format (HEX?) b'\x921\y23+bW\u68\xlb'

I could then use it in the PKCS11SignatureConfig to sign.
